### PR TITLE
Feature/milestone 6

### PR DESCRIPTION
**Pull request by oscardelalanza**

### CODE REVIEW

**Review by adriaanbd — CHANGES_REQUESTED** [review]
Please make sure that we are not able to create the same friendship twice, even though the user and friend relationship is inverted. For example, take a look at the console output below:

```
=> #<ActiveRecord::Relation [#<Friendship user_id: 1, friend_id: 2, created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">, #<Friendship user_id: 1, friend_id: 3, created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">, #<Friendship user_id: 3, friend_id: 1, created_at: "2019-12-05 20:20:13", updated_at: "2019-12-05 20:20:13">]>
```

The record pertaining to `Friendship.third` shouldn't have been able to be committed to the database and should've rolled back.

Furthermore, you can see that `user_id: 1` and `friend_id: 2` already have a friendship, but in the `user#index` view I can still see the option to add `janedoe` as a friend:

image

User janedoe:

```
=> #<User id: 2, name: "janedoe", image: nil, email: "janedoe@example.com", created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">
``` 

In addition, how are you handling friend requests? I've taken a look in your Friendship table and there is no `status` or `confirmed` column. As you know, we can't have `Friendship` record unless a Friend Request was approved from beforehand.

In regards to your tests, I'm not able to run them because it exits with an error:
```
WARNING: Rails was not able to disable referential integrity.:00,  ETA: ??:??:??

This is most likely caused due to missing permissions.
Rails needs superuser privileges to disable referential integrity.

    cause: PG::InsufficientPrivilege: ERROR:  permission denied: "RI_ConstraintTrigger_a_50952" is a system trigger
```

Nevertheless, taking a look at them and checking to see if you've tested for invalid inverted friendship, I noticed that you did:

```test 'invalid friendship request' do
    request = Friendship.new(friend_id: @friend.id)
    assert_not request.valid?
    request = Friendship.new(user_id: @user.id)
    assert_not request.valid?
  end
```

But shouldn't it have been:

```
test 'invalid friendship request' do
    request = Friendship.new(user_id: @friend.id, friend_id: @user.id)
    assert_not request.valid?
  end
```

You'd want to test the inverted relationship but I noticed that it is the same relationship. Also, does `request.valid?` commit to a database? You'd want to make sure the record is committed first before testing to see if the new inverted relationship is valid or if it could be committed to the database (it shouldn't be able too).

Please guarantee the uniqueness of user/friend record in a Friendship and make sure you are testing it properly.

Status: `Changed Requested`
Reviewer: `Adriaan Beiertz`

**Comment by tiagoit** [comment]
> Please make sure that we are not able to create the same friendship twice, even though the user and friend relationship is inverted. For example, take a look at the console output below:
> 
> ```
> => #<ActiveRecord::Relation [#<Friendship user_id: 1, friend_id: 2, created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">, #<Friendship user_id: 1, friend_id: 3, created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">, #<Friendship user_id: 3, friend_id: 1, created_at: "2019-12-05 20:20:13", updated_at: "2019-12-05 20:20:13">]>
> ```
> 
> The record pertaining to `Friendship.third` shouldn't have been able to be committed to the database and should've rolled back.
> 
> Furthermore, you can see that `user_id: 1` and `friend_id: 2` already have a friendship, but in the `user#index` view I can still see the option to add `janedoe` as a friend:
> 
> image
> 
> User janedoe:
> 
> ```
> => #<User id: 2, name: "janedoe", image: nil, email: "janedoe@example.com", created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">
> ```
> 
> In addition, how are you handling friend requests? I've taken a look in your Friendship table and there is no `status` or `confirmed` column. As you know, we can't have `Friendship` record unless a Friend Request was approved from beforehand.
> 
> In regards to your tests, I'm not able to run them because it exits with an error:
> 
> ```
> WARNING: Rails was not able to disable referential integrity.:00,  ETA: ??:??:??
> 
> This is most likely caused due to missing permissions.
> Rails needs superuser privileges to disable referential integrity.
> 
>     cause: PG::InsufficientPrivilege: ERROR:  permission denied: "RI_ConstraintTrigger_a_50952" is a system trigger
> ```
> 
> Nevertheless, taking a look at them and checking to see if you've tested for invalid inverted friendship, I noticed that you did:
> 
> ```
>     request = Friendship.new(friend_id: @friend.id)
>     assert_not request.valid?
>     request = Friendship.new(user_id: @user.id)
>     assert_not request.valid?
>   end
> ```
> 
> But shouldn't it have been:
> 
> ```
> test 'invalid friendship request' do
>     request = Friendship.new(user_id: @friend.id, friend_id: @user.id)
>     assert_not request.valid?
>   end
> ```
> 
> You'd want to test the inverted relationship but I noticed that it is the same relationship. Also, does `request.valid?` commit to a database? You'd want to make sure the record is committed first before testing to see if the new inverted relationship is valid or if it could be committed to the database (it shouldn't be able too).
> 
> Please guarantee the uniqueness of user/friend record in a Friendship and make sure you are testing it properly.
> 
> Status: `Changed Requested`
> Reviewer: `Adriaan Beiertz`



> Please make sure that we are not able to create the same friendship twice, even though the user and friend relationship is inverted. For example, take a look at the console output below:
> 
> ```
> => #<ActiveRecord::Relation [#<Friendship user_id: 1, friend_id: 2, created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">, #<Friendship user_id: 1, friend_id: 3, created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">, #<Friendship user_id: 3, friend_id: 1, created_at: "2019-12-05 20:20:13", updated_at: "2019-12-05 20:20:13">]>
> ```
> 
> The record pertaining to `Friendship.third` shouldn't have been able to be committed to the database and should've rolled back.
> 
> Furthermore, you can see that `user_id: 1` and `friend_id: 2` already have a friendship, but in the `user#index` view I can still see the option to add `janedoe` as a friend:
> 
> image
> 
> User janedoe:
> 
> ```
> => #<User id: 2, name: "janedoe", image: nil, email: "janedoe@example.com", created_at: "2019-12-05 20:19:03", updated_at: "2019-12-05 20:19:03">
> ```
> 
> In addition, how are you handling friend requests? I've taken a look in your Friendship table and there is no `status` or `confirmed` column. As you know, we can't have `Friendship` record unless a Friend Request was approved from beforehand.
> 
> In regards to your tests, I'm not able to run them because it exits with an error:
> 
> ```
> WARNING: Rails was not able to disable referential integrity.:00,  ETA: ??:??:??
> 
> This is most likely caused due to missing permissions.
> Rails needs superuser privileges to disable referential integrity.
> 
>     cause: PG::InsufficientPrivilege: ERROR:  permission denied: "RI_ConstraintTrigger_a_50952" is a system trigger
> ```
> 
> Nevertheless, taking a look at them and checking to see if you've tested for invalid inverted friendship, I noticed that you did:
> 
> ```
>     request = Friendship.new(friend_id: @friend.id)
>     assert_not request.valid?
>     request = Friendship.new(user_id: @user.id)
>     assert_not request.valid?
>   end
> ```
> 
> But shouldn't it have been:
> 
> ```
> test 'invalid friendship request' do
>     request = Friendship.new(user_id: @friend.id, friend_id: @user.id)
>     assert_not request.valid?
>   end
> ```
> 
> You'd want to test the inverted relationship but I noticed that it is the same relationship. Also, does `request.valid?` commit to a database? You'd want to make sure the record is committed first before testing to see if the new inverted relationship is valid or if it could be committed to the database (it shouldn't be able too).
> 
> Please guarantee the uniqueness of user/friend record in a Friendship and make sure you are testing it properly.
> 
> Status: `Changed Requested`
> Reviewer: `Adriaan Beiertz`

Hello @adriaanbd, thanks for the review, but you may have got confused with the milestone 5, where we should use only one row and confirmed flag for relationships. (friendship v1).

On milestone 6 (friendship v2) the friendships are only "confirmed" when we have two rows for the relationship, according to the [tutorial](https://explainextended.com/2009/03/07/selecting-friends/) this is around twice more efficient.

image

**Comment by adriaanbd** [comment]
The purpose of creating that second friendship row is to prevent a second inverted friendship to be created. The database won't allow it because there is already a record in there pertaining to that relationship; therefore, the commit would rollback. At the moment, I was able to create it from the console. Of course, if you don't agree with my review, you are free to request another code review with the same code or to raise this issue with Emily or someone at Student Success.

**Comment by ferreirati** [comment]
According to the requirements, the second row is intended to be created, it
means that the friendship is confirmed.

[image: image.png]
Let's say user_one requested friendship to user_two (1 row inserted 1->2).
When user_2 accepts the friendship request the second row is inserted
(2->1) meaning that they are now friends. We also don't need a confirmed
column in this case. See the attached image with the expected tables model.

[image: image.png]

Another benefit of having two rows for friendship is having not symmetric
relations.

[image: image.png]

I would be happy to talk more about this, we could also schedule a call for
it.

*Best Regards*

*Tiago Ferreira*
*Software Engineer*
* +55 31 98789 9961 <%2B55%2031%208789%209961>*

**Comment by tiagoit** [comment]
> The purpose of creating that second friendship row is to prevent a second inverted friendship to be created. The database won't allow it because there is already a record in there pertaining to that relationship; therefore, the commit would rollback. At the moment, I was able to create it from the console. Of course, if you don't agree with my review, you are free to request another code review with the same code or to raise this issue with Emily or someone at Student Success.

According to the requirements, the second row is intended to be created, it
means that the friendship is confirmed.

image

Let's say user_one requested friendship to user_two (1 row inserted 1->2).
When user_2 accepts the friendship request the second row is inserted
(2->1) meaning that they are now friends. We also don't need a confirmed
column in this case. See the attached image with the expected tables model.

image

Another benefit of having two rows for friendship is having not symmetric
relations.

image

I would be happy to talk more about this, we could also schedule a call for
it.

Best Regards

**Comment by adriaanbd** [comment]
I'm not talking about the `confirmed` column. I'm making reference to the fact I was able to create the same friendship twice: one with user 1 / friend 2 and another with user 2 / friend 1. However you'd like to code it that's up to you guys, but the expected behaviour is that the database shouldn't allow this. And it shouldn't allow it because you have already automatically added that second row in there when the friendship was created or through some other way. 

I'd be willing to talk about it if you'd like, please schedule a call here: www.calendly.com/adriaanbd/30min. However, please note that the issue is that the database is allowing me to create the first  user 1 / friend 2 friendship and also the second user 2 / friend 1 friendship.

**Comment by adriaanbd** [comment]
Another issue was your tests, it didn't seem you were testing it correctly, but you haven't addressed this in your comments.

**Comment by adriaanbd** [comment]
The third issue was in the UI wherein `janedoe` was displayed as an option to request friendship to but the friendship already existed. Please also address this in your comments.

**Comment by adriaanbd** [comment]
Please address issues separately, to have different discussions.

**Comment by oscardelalanza** [comment]
> I'm not talking about the `confirmed` column. I'm making reference to the fact I was able to create the same friendship twice: one with user 1 / friend 2 and another with user 2 / friend 1. However you'd like to code it that's up to you guys, but the expected behaviour is that the database shouldn't allow this. And it shouldn't allow it because you have already automatically added that second row in there when the friendship was created or through some other way.
> 
> I'd be willing to talk about it if you'd like, please schedule a call here: [www.calendly.com/adriaanbd/30min](http://www.calendly.com/adriaanbd/30min). However, please note that the issue is that the database is allowing me to create the first user 1 / friend 2 friendship and also the second user 2 / friend 1 friendship.

That is not supposed to be the mutual relationship with two rows? at least that is what the description from milestone 6 says
image

**Comment by tiagoit** [comment]
I've tested again the UI, it really had a bug on user#pending_friend_approval method.
Thanks for this.

IDK if we are making any progress on the discussion here, but I really think this is a great one, what if we all schedule a call? @oscardelalanza maybe we can share with Adriaan that video from our session...

File: app/models/user.rb [hunk 1]
```diff
@@ -23,27 +23,25 @@ class User < ApplicationRecord
...
-    friends_array += inverse_friendships.map { |friendship| friendship.user_id if friendship.confirmed }
-    friends_array.compact
+    friendships.map(&:user_id) & inverse_friendships.map(&:user_id)
   end
 
   # Requests made TO current_user
   def pending_acceptance
-    inverse_friendships.map { |friendship| friendship.user_id unless friendship.confirmed }.compact
+    inverse_friendships.map(&:user_id) - friendships.map(&:user_id)
   end
```
Comment (Torres-ssf) [inline app/models/user.rb]: I notice that you choose to not implement the 'confirmed' attribute into the friendship table. This is suggested in the first Milestone:
Screenshot from 2019-12-06 10-16-28

Implementing the status/confirmed attribute would make easy to query requests. You can do that only with Active record, without using any ruby method.

- To get pending requests:

```ruby
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```
- To get confirmed friendships:
```ruby
has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

Please take a time to read this [article](https://explainextended.com/2009/03/07/selecting-friends/) and this StackOverflow [question](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table)

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Changes required :muscle: 
Hi there team :hand: 
Great job so far :clap: 
But we still have a bug in the UI:

- [x] Even after accepting the friend request, the user index page continue to show the friend request:
Screenshot from 2019-12-06 10-12-14

**[Friendly Remainder]**
**It's best practice to add a summary of the changes in the PR. This will make your project looks more professional and by doing that, it increases the chance of you getting hired in the future. Please take the time to read this [link](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243642/). Please kindly edit the description of this PR and give it a summary like the one you saw in the article.**

Please submit for code review again after fixing this error
Happy coding :computer: 

**Sergio Torres**

File: app/models/user.rb (hunk 1 above)
Comment (tiagoit) [reply app/models/user.rb]: Fixed, we still choose not to use the confirmed flag since it's a tip from milestone 5, and milestone 6 tip is to not use it.

**Review by efrapp — APPROVED** [review]
## Hi @ferreirati and @oscardelalanza 👋

Hey guys, great job applying the suggestion from the last code review. The bug was fixed so you can merge this PR and move on! 👍

I want to share with you some suggestions that could be helpful in the future. The way you handled the  friendship flow was great, very concise code and a clever solution but bear in mind that you chose the object processing approach (using `map` etc) instead of taking advantage of the SQL queries generated by the associations (using `:through` and `:source`) which is faster.
Finally, I saw you use environment variables in your `database.yml` file for the `development` environment but think that likely you will be working in a team and there are some things that should be fast to test so in development environments it is ok to use plain information for faster testing.

That's all from me. Have a great day!

**Status:** `Approved`

**Pull request merged**